lib/sqlalchemy/sql/functions.py

```diff
@@ -440,7 +478,12 @@ def __call__(self, *c, **kwargs):
             package = None
 
         if package is not None:
-            func = _registry[package].get(fname)
+            if (
+                len(_case_sensitive_reg[package].get(fname.lower(), [])) > 1
```

Comment:
**mike bayer** (zzzeek) wrote:

does this have to be len > 1 ?  this is a more critical spot.

can we simply look in _registry[package] for the exact name match first, and then if not found fall back to lower()?  because the majority of use cases will be all-lowercase use and there should be no second check:

  reg = _registry[package]
  func = reg.get(fname, None)
  if func is None:
      func = reg.get(fname.lower())
  if func is not None:
       # ...

let me know if that works

Reply:
Here again, the `len()` thing is just to detect if this identifier is in case-sensitive or case-insensitive mode. If we always get `fname` first, then we break things.

For example: if `myfunc` and `MyFunc` are registered (so the identifier `myfunc` is switched to case-sensitive mode and is associated with the 2 case-sensitive identifiers `myfunc` and `MyFunc`), then `func.MYFUNC` will actually return the function registered with the `myfunc` case-sensitive identifier instead of an unregistered `Function`.